pkg/identity/cidr.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/labels/cidr"
+)
+
+// AllocateCIDRIdentities allocates identities for each of the specified CIDR
+// prefixes so they will be available later during policy resolution.
+//
+// On success, the returned slice will have a 1-to-1 correspondence with the
+// CIDR in the provided prefixes slice, eg the identity for prefixes[0] will
+// be held in identities[0].
+// On failure, attempts to clean up after itself and returns the error.
+func AllocateCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	if len(prefixes) == 0 {
+		return nil, nil
+	}
+
+	log.Debugf("Attempting to allocate identities for %d prefixes", len(prefixes))
+
+	// Allocate identities for each CIDR.
+	res = make([]*Identity, 0, len(prefixes))
+	for _, prefix := range prefixes {
+		if prefix == nil {
+			continue
+		}
+		labels := cidr.GetCIDRLabels(prefix)
+		id, _, err := AllocateIdentity(labels)
+		if err != nil {
+			// If any identity allocation failed, release existing identities
+			// and log the error.
+			if err2 := ReleaseSlice(res); err2 != nil {
+				log.WithError(err2).Error("Could not recover from error during CIDR identity allocation")
+			}
+			res = nil
+			break
+		}
+		res = append(res, id)
+	}
+
+	if err == nil {
+		log.Debugf("Allocated identities for %d prefixes", len(res))
+	} else {
+		log.WithError(err).Warningf("Failed to allocate identities for %d prefixes", len(res))
+	}
+
+	return res, err
```

Comment:
It might be worth returning which identities could not be allocated? I imagine the causes here could be more varied and may be specific to the identity? (I can't think of anything, though, so maybe not)

Reply:
I suppose that if we exit out & try to recover / clean up on the first failure, we could return the prefix and/or labels for that first failure, then log these in the caller. Then if there really is some problem of that variety, it should be easy to tell.

Reply:
your call. I think I came to this from the perspective of overlapping ranges sharing some identities. You corrected me above, so that leaves classes of errors that are likely (as you said) going to repeat for all of them or none of them.